armi/reactor/composites.py

```diff
@@ -771,66 +778,80 @@ def _deriveUndefinedVolume(self):
-                )
-            left.setVolume(derivedVolume)
-            totalVolume += derivedVolume
-        elif leftover:
-            runLog.warning(
-                "Gluttony Error incoming, total components {}"
-                "".format(self.getChildren())
-            )
-            for c in self.getChildren():
-                runLog.warning(
-                    "volume {} params {} for component {}"
-                    "".format(c.getVolume(), c.p, c.getName())
-                )
-            runLog.error(
-                "These are leftovers: {0}\n"
-                " Cannot deduce area of more than one component".format(leftover)
-            )
-            raise RuntimeError("Gluttony Error too many leftovers.")
-        return derivedVolume
+        # Determine the number of `DerivedShape` components and non-derived shape components
+        derivedShapeComps = [c for c in self.getChildren() if c.__class__ is components.DerivedShape]
+        otherComps = [c for c in self.getChildren() if c not in derivedShapeComps]
+
+        otherCompAreas = sum([c.getArea() for c in otherComps])
+        otherCompVolumes = sum([c.getVolume() for c in otherComps])
+
+        remainingArea = self.getMaxArea() - otherCompAreas
+        remainingVolume = self.getMaxVolume() - otherCompVolumes
+
+        geometryError = False
+        # Check for negative area
+        if remainingArea < 0:
+            geometryError = True
+            componentAreas = "\n".join(
+            [
+                f"{c:40s}, {c.getArea():.5e} cm^2"
+                for c in otherComps
+            ])
+            msg = (f"The component areas in {self} exceed the maximum "
+                   f"allowable area based on the geometry. Check that the "
+                   f"geometry is defined correctly.\n"
+                   f"Maximum allowable area: {self.getMaxArea()} cm^2\n"
+                   f"Area of all defined components: {otherCompAreas} cm^2\n"
+                   f"Breakdown: {componentAreas}")
+            runLog.error(msg)
+
+        # Check for negative volume
+        if remainingVolume < 0:
+            geometryError = True
+            componentVols = "\n".join(
+            [
+                f"{c:40s}, {c.getVolume():.5e} cc"
+                for c in otherComps
+            ])
+            msg = (f"The component volumes in {self} exceed the maximum "
+                   f"allowable volume based on the geometry. Check that the "
+                   f"geometry is defined correctly.\n"
+                   f"Maximum allowable volume: {self.getMaxVolume()} cc\n"
+                   f"Volume of all defined components: {otherCompVolumes} cc\n"
+                   f"Breakdown: {componentVols}")
+            runLog.error(msg)
+
+        if geometryError:
+            raise ValueError(f"Negative area/volume errors occurred for {self}. "
+                             "Check log for errors.")
+
```

Comment:
Handle the error if the number of derived shapes != 0 up here, so that the happy path doesnt need to be in a branch, and so that it matches the error-handling pattern above

Reply:
Moved up.